### PR TITLE
3. Require network names in cached state disk names

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

As part of ticket #4153, we want to match cached states with the correct version and network.

In PR #4391, we fixed a bug where cached state disks were missing the network name.
Now that PR has merged to `main`, and its state is rebuilt, we can require the network in disk names.

Depends-On: #4385

## Solution

Search for the exact network name in cached state disk names.

## Review

Anyone can review this PR, it's not urgent.

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: .github/workflows/continous-integration-docker.yml [hunk 1]
```diff
@@ -205,10 +206,13 @@ jobs:
       height_grep_text: 'flushing database to disk height=Height'
 
   # Test that Zebra syncs and fully validates a few thousand blocks from a cached mandatory checkpoint disk
+  #
+  # If the state version has changed, waits for the new cached state to be created.
+  # Otherwise, if the state rebuild was skipped, runs immediately after the build job.
```
Comment (dconnolly) [inline .github/workflows/continous-integration-docker.yml]: 👍

File: .github/workflows/deploy-gcp-tests.yml [hunk 2]
```diff
@@ -6,6 +6,7 @@ on:
       network:
         required: false
         type: string
+        default: Mainnet
```
Comment (dconnolly) [inline .github/workflows/deploy-gcp-tests.yml]: 👍

File: .github/workflows/deploy-gcp-tests.yml [hunk 3]
```diff
@@ -45,7 +46,6 @@ on:
         type: string
 
 env:
-  NETWORK: Mainnet
```
Comment (dconnolly) [inline .github/workflows/deploy-gcp-tests.yml]: 👍

File: .github/workflows/deploy-gcp-tests.yml [hunk 4]
```diff
@@ -196,26 +196,44 @@ jobs:
...
-      # or at least that the local state version is greater than the available cached state version from main.
+      # Find a cached state disk for this job, matching all of:
+      # - disk kind (disk_prefix) - zebra or lwd
+      # - state version (from the source code) - v{N}
+      # - network (network) - mainnet or testnet
+      # - disk target height kind (disk_suffix) - checkpoint or tip
+      #
+      # If there are multiple disks:
+      # - prefer images generated from the `main` branch, then any other branch
+      # - prefer newer images to older images
```
Comment (dconnolly) [inline .github/workflows/deploy-gcp-tests.yml]: 👍

File: .github/workflows/deploy-gcp-tests.yml [hunk 5]
```diff
@@ -196,26 +196,44 @@ jobs:
...
       #
-      # Aftwards, get the disk name to be used on further steps
-      - name: Validate local state version with cached state
+      # Passes the disk name to subsequent steps using an environmental variable.
```
Comment (dconnolly) [inline .github/workflows/deploy-gcp-tests.yml]: 👍

File: .github/workflows/deploy-gcp-tests.yml [hunk 6]
```diff
@@ -196,26 +196,44 @@ jobs:
...
+              CACHED_DISK_NAME=$(gcloud compute images list --verbosity=debug --filter="name~${{ inputs.disk_prefix }}-.+-[0-9a-f]+-v${LOCAL_STATE_VERSION}-${NETWORK}-${{ inputs.disk_suffix }}" --format="value(NAME)" --sort-by=~creationTimestamp --limit=1)
+              echo "Disk: $CACHED_DISK_NAME"
+          fi
+
+          if [[ -z "$CACHED_DISK_NAME" ]]; then
+              echo "No cached state disk available"
+              echo "Expected ${{ inputs.disk_prefix }}-(branch)-[0-9a-f]+-v${LOCAL_STATE_VERSION}-${NETWORK}-${{ inputs.disk_suffix }}"
+              echo "Cached state test jobs must depend on the cached state rebuild job"
+              exit 1
+          fi
```
Comment (dconnolly) [inline .github/workflows/deploy-gcp-tests.yml]: 👍

**Review by dconnolly — DISMISSED** [review]

**Comment by dconnolly** [comment]
@Mergifyio update

**Review by gustavovalverde — APPROVED** [review]
🚀

File: .github/workflows/deploy-gcp-tests.yml [hunk 7]
```diff
@@ -241,24 +241,18 @@ jobs:
 
           # Try to find an image generated from the main branch
           # Fields are listed in the "Create image from state disk" step
-          #
-          # TODO: require ${NETWORK} in the name after PR #4391 merges to main, and runs a full sync
-          #       network should replace [a-z]*
-          CACHED_DISK_NAME=$(gcloud compute images list --filter="name~${DISK_PREFIX}-main-[0-9a-f]+-v${LOCAL_STATE_VERSION}-[a-z]*-${{ inputs.disk_suffix }}" --format="value(NAME)" --sort-by=~creationTimestamp --limit=1)
+          CACHED_DISK_NAME=$(gcloud compute images list --filter="name~${DISK_PREFIX}-main-[0-9a-f]+-v${LOCAL_STATE_VERSION}-${NETWORK}-${{ inputs.disk_suffix }}" --format="value(NAME)" --sort-by=~creationTimestamp --limit=1)
```
Comment (teor2345) [inline .github/workflows/deploy-gcp-tests.yml:244]: This code is partly duplicated in `get-available-disks` in PR #4385, so this PR also needs to add the network to that file.

Comment (gustavovalverde) [reply .github/workflows/deploy-gcp-tests.yml:244]: Done

**Review by gustavovalverde — APPROVED** [review]
Suggestions applied

**Pull request merged**